presto-plan-checker-providers/src/main/java/com/facebook/presto/plancheckerproviders/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,182 @@
+    private static final String PRESTO_CATALOG_PROPERTY = "X-Presto-Catalog-Property";
+    private static final MediaType JSON_CONTENT_TYPE = MediaType.parse("application/json; charset=utf-8");
+    public static final String PLAN_CONVERSION_ENDPOINT = "/v1/velox/plan";
+
+    private final NodeManager nodeManager;
+    private final JsonCodec<SimplePlanFragment> planFragmentJsonCodec;
+    private final NativePlanCheckerConfig config;
+    private final OkHttpClient httpClient;
+
+    public NativePlanChecker(NodeManager nodeManager, JsonCodec<SimplePlanFragment> planFragmentJsonCodec, NativePlanCheckerConfig config)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.planFragmentJsonCodec = requireNonNull(planFragmentJsonCodec, "planFragmentJsonCodec is null");
+        this.config = requireNonNull(config, "config is null");
+        this.httpClient = new OkHttpClient.Builder().build();
+    }
+
+    private URL getPlanValidateUrl()
+    {
+        Set<Node> nodes = nodeManager.getAllNodes();
+        Node sidecarNode = nodes.stream().filter(Node::isCoordinatorSidecar).findFirst()
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, "could not find native sidecar node"));
+        try {
+            // endpoint to retrieve session properties from native worker
+            return new URL(sidecarNode.getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT);
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        try {
+            if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
+                runValidation(planFragment);
+            }
+            else if (LOG.isDebugEnabled()) {
+                LOG.debug("Skipping Native Plan Validation for plan fragment id: %s", planFragment.getId());
+            }
+        }
+        catch (final PrestoException e) {
+            if (config.isQueryFailOnError()) {
+                throw e;
+            }
+        }
+        catch (final Exception e) {
+            LOG.error(e, "Failed to run native plan validation");
+            if (config.isQueryFailOnError()) {
+                throw new IllegalStateException("Failed to run native plan validation", e);
+            }
+        }
+    }
+
+    private boolean isInternalSystemConnector(PlanNode planNode)
+    {
+        try {
+            return planNode.accept(new CheckInternalVisitor(), null);
+        }
+        catch (final IllegalArgumentException e) {
+            // An InternalPlanNode will not allow a `PlanVisitor`, allow it to pass
+            return false;
```

Comment:
This was because of the check in `InternalPlanVisitor` here https://github.com/prestodb/presto/blob/e6470a934f21d8a69fbeef36f24af8c4d8706d1e/presto-main/src/main/java/com/facebook/presto/sql/planner/plan/InternalPlanNode.java#L35

I don't know another way around that, are you saying an `InternalPlanNode` can still get sent to a native worker?

Reply:
> I don't know another way around that, are you saying an InternalPlanNode can still get sent to a native worker?

yes, InternalPlanNode can get sent to native workers. 

I think probably I need to understand why this check is here, and then can help come up with alternatives.

Reply:
Much of the same system connector queries also have PlanFragments with a node that extends `InternalPlanNode`
For example, the query `SELECT table_name FROM information_schema.tables WHERE table_schema = 'tpch';` has an `ExchangeNode`, here is debug info of the node
```
planNode = {ProjectNode@19451} 
 source = {ExchangeNode@19697} 
  sourceLocation = {Optional@19700} "Optional.empty"
  type = {ExchangeNode$Type@19705} "REPARTITION"
  scope = {ExchangeNode$Scope@19706} "LOCAL"
  sources = {SingletonImmutableList@19707}  size = 1
  partitioningScheme = {PartitioningScheme@19708} "ConnectorTableMetadata{partitioning=Partitioning{handle=ROUND_ROBIN, arguments=[]}, outputLayout=[field, field_0, field_1, field_2, field_3, field_4, field_5, field_6, field_7, field_8], hashChannel=Optional.empty, replicateNullsAndAny=false, bucketToPartition=Optional.empty}"
  inputs = {SingletonImmutableList@19709}  size = 1
  ensureSourceOrdering = false
  orderingScheme = {Optional@19700} "Optional.empty"
  id = {PlanNodeId@19710} "327"
  statsEquivalentPlanNode = {Optional@19700} "Optional.empty"
 assignments = {Assignments@19698} 
 locality = {ProjectNode$Locality@19699} "LOCAL"
```
and results in the error
```
"java.lang.IllegalArgumentException: PlanVisitor is only for connector to use; InternalPlanNode should never use it"
 backtrace = {Object[5]@19470} 
 detailMessage = "PlanVisitor is only for connector to use; InternalPlanNode should never use it"
 cause = {IllegalArgumentException@19452} "java.lang.IllegalArgumentException: PlanVisitor is only for connector to use; InternalPlanNode should never use it"
 stackTrace = {StackTraceElement[93]@19476} 
  0 = {StackTraceElement@19511} "com.google.common.base.Preconditions.checkArgument(Preconditions.java:143)"
  1 = {StackTraceElement@19512} "com.facebook.presto.sql.planner.plan.InternalPlanNode.accept(InternalPlanNode.java:35)"
  2 = {StackTraceElement@19513} "com.facebook.presto.plancheckerproviders.nativechecker.NativePlanChecker$CheckInternalVisitor.visitPlan(NativePlanChecker.java:152)"
  3 = {StackTraceElement@19514} "com.facebook.presto.plancheckerproviders.nativechecker.NativePlanChecker$CheckInternalVisitor.visitPlan(NativePlanChecker.java:138)"
  4 = {StackTraceElement@19515} "com.facebook.presto.spi.plan.PlanVisitor.visitProject(PlanVisitor.java:35)"
  5 = {StackTraceElement@19516} "com.facebook.presto.spi.plan.ProjectNode.accept(ProjectNode.java:108)"
  6 = {StackTraceElement@19517} "com.facebook.presto.plancheckerproviders.nativechecker.NativePlanChecker.isInternalSystemConnector(NativePlanChecker.java:98)"
```

So ignoring the exception during the `isInternalSystemConnector()` seemed to be the only way to handle these type of PlanFragments

Reply:
as per the other comment, I think we should get to the root cause of the system connector issue rather than taking this visitor approach.

Reply:
I think if we separate out the InternalPlanNode/PlanNode issue, we'll still have problems with system connectors because the system connector isn't implemented in C++, and the plan converter will through when it encounters them.  This problem has been encountered in other contexts too (e.g. #21725, #21285).